core/consensus/babe/src/lib.rs

```diff
@@ -1195,6 +1197,117 @@ pub fn import_queue<B, E, Block: BlockT<Hash=H256>, I, RA, PRA>(
+	C::Api: BabeApi<<C as BlockOf>::Type>
+{
+	client
+		.cache()
+		.and_then(|cache| cache
+			.get_at(&well_known_cache_keys::AUTHORITIES, at)
+			.and_then(|v| Decode::decode(&mut &v[..]))
+		)
+		.or_else(|| BabeApi::epoch(&*client.runtime_api(), at).ok()
+			.map(|epoch| epoch.authorities.into_iter().map(|x| x.0).collect())
+		)
+		.ok_or_else(|| consensus_common::Error::InvalidAuthoritiesSet.into())
+}
+
+/// Provide the authority key, if any, that is controlled by this node as of the given block.
+fn authority<C>(client: &C, keystore: Arc<Store>) -> Option<AuthorityPair> where
+	C: ProvideRuntimeApi + BlockOf + ProvideCache<<C as BlockOf>::Type>
+	+ HeaderBackend<<C as BlockOf>::Type>,
+	C::Api: BabeApi<<C as BlockOf>::Type>
+{
+	let owned = keystore.contents::<AuthorityId>().ok()?;
+	let at = BlockId::Number(client.info().best_number);
+	/// The list of authority keys that is current. By default this will just use the state of
+	/// the best block, but you might want it to use some other block's state instead if it's
+	/// more sophisticated. Grandpa, for example, will probably want to use the state of the last
+	/// finalised block.
+	let authorities = authorities_at::<C>(client, &at).ok()?;
+	let maybe_pub = owned.into_iter()
+		.find(|i| authorities.contains(i));
+	maybe_pub.and_then(|public| keystore.load(&public, "").ok())
+}
+
+/// Type of source for block sealing. Different consensus algorithms have different sealing
+/// methods; for PoW it'll be a miner or mining instance. For PoA/PoS it'll be a key type.
+pub trait SealingSource {
+	/// Human-readable description of this general class.
+	///
+	/// e.g. `"ProgPoW"`, `"S/R 25519 Key"`.
+	const SEALER_TYPE: &'static str;
+	/// Human-readable description of this specific instance.
+	///
+	/// e.g. `"OpenCL GPU Radeon 8970 @ slot 1"`, `"5Gav1mbbo348grBREHTYgevvf4Gfe5GFE4"`.
+	fn format(&self) -> String;
+}
+
+impl<T: primitives::crypto::AppPublic> SealingSource for T {
+	const SEALER_TYPE: &'static str = "Key";
+	fn format(&self) -> String { use primitives::crypto::Ss58Codec; self.to_ss58check() }
+}
+
+/// A runtime service for a consensus worker.
+///
+/// This is administered by the client service to help it spin up everything necessary. It can
+/// provide API hooks for handling user-level/RPC events. It can return information regarding
+/// the status of the service.
+pub trait Service<'a>: 'a {
+	/// Instance of the source of sealing. Different consensus algorithms have different sealing
+	/// methods; for PoW it'll be a miner or mining instance. For PoA/PoS it'll be a key type.
+	type Sealer: SealingSource;
+
+	/// The Client type. Different services can make different constraints on this.
+	type Client: ProvideRuntimeApi + BlockOf + ProvideCache<<Self::Client as BlockOf>::Type>
+		+ HeaderBackend<<Self::Client as BlockOf>::Type>;
+
+	/// Initialize the consensus service.
+	fn initialize(client: &'a Self::Client, keystore: Arc<Store>) -> Self;
```

Comment:
There doesn't seem to be a good reason why we should have service objects at all. 

The design of the pluggable consensus system was intended to avoid over-assumption about what resources a consensus system might need access to, and I'm afraid that's exactly what this trait function is -- it's omitting at a bare minimum network and event-loop access.

The primitive of "some thing that can be driven to completion" (i.e. a `Future`) is much less fragile.
I understand the motivation of wanting to avoid users having to write any consensus-specific initialization in their service files, but it should be 1 line per consensus engine at most, if we instead go the route of consensus crates exposing an `initialize` function (without fixed type signature!). 

That is much more anti-fragile compared to a bunch of trait machinery/macro magic trying to create an idealized declarative system that will almost certainly have to be ripped up later.

Reply:
Additionally, considering that this PR is pressing, let's please not introduce API changes that aren't strictly necessary, because we should discuss such API changes beforehand.

Reply:
My concern is that it *actually* is "1 line per consensus engine at most".

Right now it's a load of incoherent boilerplate that chain creators can neither write nor understand.